File: src/main/java/org/trostheide/generativeArt/LSystemGenerator.java
package org.trostheide.generativeArt;

import org.trostheide.generativeArt.core.ArtGenerator;
import org.trostheide.generativeArt.core.ParameterDefinition;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Stack;

public class LSystemGenerator implements ArtGenerator {

    @Override
    public String getId() {
        return "l-system";
    }

    @Override
    public String getDisplayName() {
        return "L-System Fractal";
    }

    @Override
    public List<ParameterDefinition> getParameterDefinitions() {
        return List.of(
                ParameterDefinition.selection("Preset", "Custom", 
                    java.util.Arrays.asList("Custom", "Koch Snowflake", "Quadratic Koch", "Sierpinski", "Dragon Curve", "Plant/Bush", "Fern", "Hilbert Curve", "Penrose Tiling"), 
                    "Select a predefined L-System"),
                ParameterDefinition.integer("Iterations", 4, 1, 12, "Number of recursion steps"),
                ParameterDefinition.doubleVal("Angle", 90.0, 0.0, 360.0, "Turn angle in degrees"),
                ParameterDefinition.doubleVal("Line Length", 10.0, 1.0, 100.0, "Length of each segment"),
                ParameterDefinition.string("Axiom", "F", "Initial state (e.g., F)"),
                ParameterDefinition.string("Rules", "F:F+F-F-F+F", "Production rules (e.g., F:F+F-F-F+F)"));
    }

    @Override
    public boolean onParameterChanged(String paramName, Object newValue, Map<String, Object> currentValues) {
        if ("Preset".equals(paramName) && newValue instanceof String) {
            String preset = (String) newValue;
            switch (preset) {
                case "Koch Snowflake":
                    currentValues.put("Axiom", "F++F++F");
                    currentValues.put("Rules", "F:F-F++F-F");
                    currentValues.put("Angle", 60.0);
                    currentValues.put("Iterations", 4);
                    return true;
                case "Quadratic Koch":
                    currentValues.put("Axiom", "F");
                    currentValues.put("Rules", "F:F+F-F-F+F");
                    currentValues.put("Angle", 90.0);
                    currentValues.put("Iterations", 4);
                    return true;
                case "Sierpinski":
                    currentValues.put("Axiom", "F-G-G");
                    currentValues.put("Rules", "F:F-G+F+G-F;G:GG");
                    currentValues.put("Angle", 120.0);
                    currentValues.put("Iterations", 6);
                    return true;
                case "Dragon Curve":
                    currentValues.put("Axiom", "FX");
                    currentValues.put("Rules", "X:X+YF+;Y:-FX-Y");
                    currentValues.put("Angle", 90.0);
                    currentValues.put("Iterations", 12);
                    return true;
                case "Plant/Bush":
                    currentValues.put("Axiom", "F");
                    currentValues.put("Rules", "F:FF+[+F-F-F]-[-F+F+F]");
                    currentValues.put("Angle", 22.0);
                    currentValues.put("Iterations", 5);
                    return true;
                case "Fern":
                    currentValues.put("Axiom", "X");
                    currentValues.put("Rules", "X:F+[[X]-X]-F[-FX]+X;F:FF");
                    currentValues.put("Angle", 25.0);
                    currentValues.put("Iterations", 5);
                    return true;
                case "Hilbert Curve":
                    currentValues.put("Axiom", "X");
                    currentValues.put("Rules", "X:-YF+XFX+FY-;Y:+XF-YFY-FX+");
                    currentValues.put("Angle", 90.0);
                    currentValues.put("Iterations", 5);
                    return true;
                case "Penrose Tiling": // Approximated with basic LSystem capabilities
                    currentValues.put("Axiom", "[X]++[X]++[X]++[X]++[X]");
                    currentValues.put("Rules", "W:YF++ZF----XF[-YF----WF]++;X:+YF--ZF[---WF--XF]+;Y:-WF++XF[+++YF++ZF]-;Z:--YF++++WF[+ZF++++XF]--XF;F:");
                    currentValues.put("Angle", 36.0);
                    currentValues.put("Iterations", 5);
                    return true;
                case "Custom":
                default:
                    return false; // Do nothing for custom, let user edit
            }
        }
        
        // If they edit a field manually while a preset is active, we could optionally switch back to "Custom"
        if (!"Preset".equals(paramName) && !"Custom".equals(currentValues.get("Preset"))) {
            currentValues.put("Preset", "Custom");
            return true;
        }
        
        return false;
    }

    @Override
    public String generate(Map<String, Object> params) {
        // Params
        int iterations = (int) params.getOrDefault("Iterations", 4);
        double angle = (double) params.getOrDefault("Angle", 90.0);
        double length = (double) params.getOrDefault("Line Length", 10.0);
        String axiom = (String) params.getOrDefault("Axiom", "F");
        String rulesStr = (String) params.getOrDefault("Rules", "F:F+F-F-F+F");

        // Dimensions
        double width = 1000;
        double height = 1000;
        if (params.containsKey("width"))
            width = ((Number) params.get("width")).doubleValue();
        if (params.containsKey("height"))
            height = ((Number) params.get("height")).doubleValue();

        // Parse Rules
        Map<Character, String> rules = new HashMap<>();
        for (String part : rulesStr.split(";")) {
            String[] kv = part.split(":");
            if (kv.length == 2)
                rules.put(kv[0].charAt(0), kv[1]);
        }

        // Expand
        String current = axiom;
        for (int i = 0; i < iterations; i++) {
            StringBuilder sb = new StringBuilder();
            for (char c : current.toCharArray()) {
                sb.append(rules.getOrDefault(c, String.valueOf(c)));
            }
            current = sb.toString();
        }

        // Draw
        double x = 0;
        double y = 0;
        double currentAngle = 0; // facing right

        double minX = 0, maxX = 0;
        double minY = 0, maxY = 0;

        StringBuilder path = new StringBuilder();
        path.append(String.format(java.util.Locale.US, "M %.1f %.1f", x, y));

        Stack<double[]> stack = new Stack<>();

        for (char c : current.toCharArray()) {
            if (c == 'F' || c == 'G') {
                double x2 = x + Math.cos(Math.toRadians(currentAngle)) * length;
                double y2 = y + Math.sin(Math.toRadians(currentAngle)) * length;
                path.append(String.format(java.util.Locale.US, " L %.1f %.1f", x2, y2));
                x = x2;
                y = y2;
                
                if (x < minX) minX = x;
                if (x > maxX) maxX = x;
                if (y < minY) minY = y;
                if (y > maxY) maxY = y;
            } else if (c == '+') {
                currentAngle += angle;
            } else if (c == '-') {
                currentAngle -= angle;
            } else if (c == '[') {
                stack.push(new double[] { x, y, currentAngle });
            } else if (c == ']') {
                if (!stack.isEmpty()) {
                    double[] state = stack.pop();
                    x = state[0];
                    y = state[1];
                    currentAngle = state[2];
                    path.append(String.format(java.util.Locale.US, " M %.1f %.1f", x, y));
                }
            }
        }

        double margin = 50.0;
        double bboxWidth = maxX - minX;
        double bboxHeight = maxY - minY;
        
        // Prevent division by zero
        if (bboxWidth == 0) bboxWidth = 1;
        if (bboxHeight == 0) bboxHeight = 1;

        double scale = Math.min((width - 2 * margin) / bboxWidth, (height - 2 * margin) / bboxHeight);
        double offsetX = (width - bboxWidth * scale) / 2.0 - minX * scale;
        double offsetY = (height - bboxHeight * scale) / 2.0 - minY * scale;

        return String.format(java.util.Locale.US,
                "<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 %.1f %.1f' width='%.1f' height='%.1f'>\n" +
                        "<defs><clipPath id='pageClip'><rect width='%.1f' height='%.1f'/></clipPath></defs>\n" +
                        "<rect width='%.1f' height='%.1f' fill='white'/>\n" +
                        "<g clip-path='url(#pageClip)'>\n" +
                        "  <g transform='translate(%.1f, %.1f) scale(%.4f)'>\n" +
                        "    <path d='%s' stroke='black' fill='none' stroke-width='1' vector-effect='non-scaling-stroke'/>\n" +
                        "  </g>\n" +
                        "</g>\n</svg>",
                width, height, width, height, width, height, width, height, offsetX, offsetY, scale, path.toString());
    }
}

File: web/src/generators/LSystemGenerator.js
import { Generator } from '../core/Generator';
import { ParameterDefinition } from '../core/ParameterDefinition';

export class LSystemGenerator extends Generator {
    getId() {
        return "l-system";
    }

    getDisplayName() {
        return "L-System Fractal";
    }

    getParameterDefinitions() {
        return [
            ParameterDefinition.selection("Preset", "Custom", 
                ["Custom", "Koch Snowflake", "Quadratic Koch", "Sierpinski", "Dragon Curve", "Plant/Bush", "Fern", "Hilbert Curve", "Penrose Tiling"], 
                "Select a predefined L-System"),
            ParameterDefinition.integer("Iterations", 4, 1, 12, "Number of recursion steps"),
            ParameterDefinition.doubleVal("Angle", 90.0, 0.0, 360.0, "Turn angle in degrees"),
            ParameterDefinition.doubleVal("Line Length", 10.0, 1.0, 100.0, "Length of each segment"),
            ParameterDefinition.string("Axiom", "F", "Initial state"),
            ParameterDefinition.string("Rules", "F:F+F-F-F+F", "Production rules")
        ];
    }

    onParameterChanged(paramName, newValue, currentValues) {
        if (paramName === "Preset" && typeof newValue === "string") {
            const preset = newValue;
            switch (preset) {
                case "Koch Snowflake":
                    currentValues["Axiom"] = "F++F++F";
                    currentValues["Rules"] = "F:F-F++F-F";
                    currentValues["Angle"] = 60.0;
                    currentValues["Iterations"] = 4;
                    return true;
                case "Quadratic Koch":
                    currentValues["Axiom"] = "F";
                    currentValues["Rules"] = "F:F+F-F-F+F";
                    currentValues["Angle"] = 90.0;
                    currentValues["Iterations"] = 4;
                    return true;
                case "Sierpinski":
                    currentValues["Axiom"] = "F-G-G";
                    currentValues["Rules"] = "F:F-G+F+G-F;G:GG";
                    currentValues["Angle"] = 120.0;
                    currentValues["Iterations"] = 6;
                    return true;
                case "Dragon Curve":
                    currentValues["Axiom"] = "FX";
                    currentValues["Rules"] = "X:X+YF+;Y:-FX-Y";
                    currentValues["Angle"] = 90.0;
                    currentValues["Iterations"] = 12;
                    return true;
                case "Plant/Bush":
                    currentValues["Axiom"] = "F";
                    currentValues["Rules"] = "F:FF+[+F-F-F]-[-F+F+F]";
                    currentValues["Angle"] = 22.0;
                    currentValues["Iterations"] = 5;
                    return true;
                case "Fern":
                    currentValues["Axiom"] = "X";
                    currentValues["Rules"] = "X:F+[[X]-X]-F[-FX]+X;F:FF";
                    currentValues["Angle"] = 25.0;
                    currentValues["Iterations"] = 5;
                    return true;
                case "Hilbert Curve":
                    currentValues["Axiom"] = "X";
                    currentValues["Rules"] = "X:-YF+XFX+FY-;Y:+XF-YFY-FX+";
                    currentValues["Angle"] = 90.0;
                    currentValues["Iterations"] = 5;
                    return true;
                case "Penrose Tiling": // Approximated with basic LSystem capabilities
                    currentValues["Axiom"] = "[X]++[X]++[X]++[X]++[X]";
                    currentValues["Rules"] = "W:YF++ZF----XF[-YF----WF]++;X:+YF--ZF[---WF--XF]+;Y:-WF++XF[+++YF++ZF]-;Z:--YF++++WF[+ZF++++XF]--XF;F:";
                    currentValues["Angle"] = 36.0;
                    currentValues["Iterations"] = 5;
                    return true;
                case "Custom":
                default:
                    return false; // Do nothing for custom, let user edit
            }
        }
        
        // Switch to custom if another field is manually changed while a preset is active
        if (paramName !== "Preset" && currentValues["Preset"] !== "Custom") {
            currentValues["Preset"] = "Custom";
            return true;
        }
        
        return false;
    }

    generate(params) {
        const iterations = params["Iterations"] || 4;
        const angle = params["Angle"] || 90.0;
        const length = params["Line Length"] || 10.0;
        const axiom = params["Axiom"] || "F";
        const rulesStr = params["Rules"] || "F:F+F-F-F+F";

        const width = params["width"] || 1000;
        const height = params["height"] || 1000;

        // Parse Rules
        const rules = {};
        rulesStr.split(';').forEach(part => {
            const [k, v] = part.split(':');
            if (k && v) rules[k] = v;
        });

        // Expand
        let current = axiom;
        for (let i = 0; i < iterations; i++) {
            let next = "";
            for (const char of current) {
                next += rules[char] || char;
            }
            current = next;
        }

        // Draw
        let x = 0;
        let y = 0;
        let currentAngle = 0; // facing right (0 deg in SVG coord system is right)

        let minX = 0, maxX = 0;
        let minY = 0, maxY = 0;

        // In Java implementation, 0 angle was implied right?
        // Java: x + cos(rad) * len.

        // Using basic path building manually instead of SvgCanvas because LSystem returns a single path usually
        // But to be consistent let's follow the standard return type.
        // The Java code returns a full Svg string directly, skipping SvgCanvas.

        let pathData = `M ${x.toFixed(1)} ${y.toFixed(1)}`;
        const stack = [];

        for (const char of current) {
            if (char === 'F' || char === 'G') {
                const rad = currentAngle * Math.PI / 180;
                const x2 = x + Math.cos(rad) * length;
                const y2 = y + Math.sin(rad) * length;
                pathData += ` L ${x2.toFixed(1)} ${y2.toFixed(1)}`;
                x = x2;
                y = y2;
                
                if (x < minX) minX = x;
                if (x > maxX) maxX = x;
                if (y < minY) minY = y;
                if (y > maxY) maxY = y;
            } else if (char === '+') {
                currentAngle += angle;
            } else if (char === '-') {
                currentAngle -= angle;
            } else if (char === '[') {
                stack.push({ x, y, currentAngle });
            } else if (char === ']') {
                if (stack.length > 0) {
                    const state = stack.pop();
                    x = state.x;
                    y = state.y;
                    currentAngle = state.currentAngle;
                    pathData += ` M ${x.toFixed(1)} ${y.toFixed(1)}`;
                }
            }
        }

        const margin = 50.0;
        let bboxWidth = maxX - minX;
        let bboxHeight = maxY - minY;

        if (bboxWidth === 0) bboxWidth = 1;
        if (bboxHeight === 0) bboxHeight = 1;

        const scale = Math.min((width - 2 * margin) / bboxWidth, (height - 2 * margin) / bboxHeight);
        const offsetX = (width - bboxWidth * scale) / 2.0 - minX * scale;
        const offsetY = (height - bboxHeight * scale) / 2.0 - minY * scale;

        // Simple SVG return
        return `<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 ${width} ${height}' width='${width}' height='${height}'>
            <rect width='${width}' height='${height}' fill='white'/>
            <g transform='translate(${offsetX.toFixed(1)}, ${offsetY.toFixed(1)}) scale(${scale.toFixed(4)})'>
                <path d='${pathData}' stroke='black' fill='none' stroke-width='1' vector-effect='non-scaling-stroke'/>
            </g>
        </svg>`;
    }
}
